### ISSUE TITLE
Unable to setup IPFS Swarm. failed to negotiate security protocol: context deadline exceeded.

### ISSUE THREAD

**Issue by AdwaitB** [ipfs/go-ipfs#6468]
#### Version information:
go-ipfs version: 0.4.22-dev-810cb607e-dirty
Repo version: 7
System version: amd64/linux
Golang version: go1.12.5

#### Description:
Unable to set up IPFS Swarm.

I am trying to deploy a n node swarm on bare metal machines on Grid5000.
I tried the same steps for n VM running on Vagrant, but I get the same error.

Did the following steps:
1. Build IPFS (See commit hash in version info)
2. `ipfs init`
3. generate swarm.key and copy it to `~/.ipfs/swarm.key` on all nodes. (I am using ipfs-swarm-key-gen)
4. `ipfs bootstrap rm --all`
5. `ipfs bootstrap add /ip4/<ip>/tcp/4001/ipfs/<node_hash>` for all nodes. Here for n nodes, I am adding the multiaddr for the remaining n-1 nodes.

`ipfs swarm peers` returns empty.
`ipfs bootstrap` gives the n-1 nodes mutiadd.
`ipfs swarm addrs` gives the  `<hash> \n /ip4/<ip>/tcp/4001` for the n-1 nodes.

I tried to telnet from one node to see if the port 4001 is accessible. It is accessible. Although, 8080 doesn't work from outside the node.
I tried adding some files from one node,  but `ipfs get <hash>` never ends.

I found the following in the debug output of the daemon
`swarm2: got error on dial: failed to negotiate security protocol: context deadline exceeded swarm_dial.go:420`

Can anyone help me with this.

Thank you

**Comment by Stebalien** [comment]
Just to check are you using the _same_ swarm key on each node?

**Comment by AdwaitB** [comment]
Yes, I was using the same swarm key. This issue was fixed if I removed the latencies that were added using tc (around 100 ms)

**Comment by Stebalien** [comment]
Ah... So, we set a really short timeout for local connections (5 seconds). Unfortunately, this applies to the entire connection process, not just the initial handshake. We do this as we limit the number of simultaneous active dials because they hold open file descriptors.

See: https://github.com/libp2p/go-libp2p/issues/666

If you want, you can try applying the following patch and rebuilding go-ipfs. That _should_ fix this:


```diff
diff --git a/cmd/ipfs/main.go b/cmd/ipfs/main.go
index 26973553d..b51917f15 100644
--- a/cmd/ipfs/main.go
+++ b/cmd/ipfs/main.go
@@ -28,12 +28,18 @@ import (
 	"github.com/ipfs/go-ipfs-config"
 	u "github.com/ipfs/go-ipfs-util"
 	logging "github.com/ipfs/go-log"
+	transport "github.com/libp2p/go-libp2p-core/transport"
 	loggables "github.com/libp2p/go-libp2p-loggables"
+	swarm "github.com/libp2p/go-libp2p-swarm"
 	ma "github.com/multiformats/go-multiaddr"
 	madns "github.com/multiformats/go-multiaddr-dns"
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+func init() {
+	swarm.DialTimeoutLocal = transport.DialTimeout
+}
+
 // log is the command logger
 var log = logging.Logger("cmd/ipfs")
 
```
